
- Автор: Ролдугин Максим
- Дата: 19.03.2026

### Цели и задачи проекта

Необходимо познакомиться с данными представленными в датасете, проанализировать их: посмотреть структуру, наличие пропусков, типы данных, корректность названий. Привести данные в достойный вид для последующей работы. Далее отобрать данные за период 2003-2013, создать новые категории, разделить на группы, выделить топ 7 по количеству.


### Описание данных

<font color='#777778'>-
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</font>

### Содержимое проекта

<font color='#777778'> Проект: Рынок видеоигр
1) Загрузить датасет и познакомиться с данными
2) Проанализировать данные, сделать выводы
3) Отметить особенности в выгруженном датасете
4) Проверить данные на ошибки (название, типы, дубликаты)
5) Фильтрация
6) Категоризация
7) Вывод топ 7 </font>



## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [96]:
import pandas as pd

In [97]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [98]:
display(df.head())

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [99]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


None

В предоставленном датасете 16955 строк и 11 столбцов. Данные соответствуют описанию.
Пропуски есть: Name - 2, Year of Release - 275, Genre - 2, Critic Score - 8714, User Score - 6804, Rating -  6871  .
Многие столбцы нужно привести к соответствующему типу.
Название столбцов имеют пробелы, нужно привести к стилю snake case.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [100]:
display(df.columns)

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [101]:
df.columns = df.columns.str.lower()

In [102]:
df.columns = df.columns.str.replace(' ', '_')
display(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [103]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce')

In [104]:
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce')

In [105]:
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce')

In [106]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


None

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [107]:
miss_absolut = df.isna().sum()
display(miss_absolut)

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [108]:
miss_otnosit = miss_absolut / len(df)
display(miss_otnosit)

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


Пропуски характерны для столбцов: name - 2, year_of_release - 275, genre - 2, eu_sales - 6, critic_score - 8714, user_score - 6804, rating - 6871

Скорее всего в name и genre пропуски появились из-за человеческого фактора. В critic_score, user_score, rating из-за технических ошибок или изначально не вписанных данных. В eu_sales из-за преобразования данных.

В name и genre можно просто удалить, в eu_sales можно заменить нулями, тк доля пропусков незначительна. В year_of_release важен для анализа, будет разумно заменить на медиану.В остальных столбцах либо ооочень много пропусков, либо их нет. Скорее всего нужно оставить как есть, так как заменить от 40 до 51 % данных на *что-то* полностью искозит их достоверность (наверное)

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [109]:
df = df.dropna(subset=['genre'])

In [110]:
df = df.dropna(subset=['name'])

In [111]:
df['rating'] = df['rating'].fillna(0)

In [112]:
df['critic_score'] = df['critic_score'].fillna(0)

In [113]:
df['user_score'] = df['user_score'].fillna(0)

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [114]:
unique_genre = df['genre'].unique()
display(unique_genre)
print(len(unique_genre))

unique_platform = df['platform'].unique()
display(unique_platform)
print(len(unique_platform))

unique_rating = df['rating'].unique()
display(unique_rating)
print(len(unique_rating))

unique_year = df['year_of_release'].unique()
display(unique_year)
print(len(unique_year))

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

24


array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

31


array(['E', 0, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

9


array([2006., 1985., 2008., 2009., 1996., 1989., 1984., 2005., 1999.,
       2007., 2010., 2013., 2004., 1990., 1988., 2002., 2001., 2011.,
       1998., 2015., 2012., 2014., 1992., 1997., 1993., 1994., 1982.,
       2016., 2003., 1986., 2000.,   nan, 1995., 1991., 1981., 1987.,
       1980., 1983.])

38


In [115]:
df['genre']=df['genre'].str.lower()

In [116]:
df['platform']=df['platform'].str.lower()

In [117]:
df['name']=df['name'].str.lower()

In [118]:
df['rating']=df['rating'].str.upper()

In [119]:
print(f'Уникальных значений после в unique_genre: {len(unique_genre)}')
print(f'Уникальных значений после в unique_platform: {len(unique_platform)}')
print(f'Уникальных значений после в unique_rating: {len(unique_rating)}')
print(f'Уникальных значений после в unique_year: {len(unique_year)}')

Уникальных значений после в unique_genre: 24
Уникальных значений после в unique_platform: 31
Уникальных значений после в unique_rating: 9
Уникальных значений после в unique_year: 38


- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [120]:
print('Явные дубликаты:', df.duplicated().sum())
df = df.drop_duplicates()

Явные дубликаты: 241


В genre найдено 12 неявныз дубликатов
Явных 241 было обнаружено.
Чтобы убрать их привел все к одному регистру, а явные просто удалил

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [121]:
initial_rows = 16956
current_rows = len(df)
deleted_rows_absolut = initial_rows - current_rows
deleted_rows_otnosit = deleted_rows_absolut / initial_rows

print(f'Начальное количество строк: {initial_rows}')
print(f'Текущее количество строк: {current_rows}')
print(f'Удалено строк (абсолютное): {deleted_rows_absolut}')
print(f'Удалено строк (относительное): {deleted_rows_otnosit:.2%}')

Начальное количество строк: 16956
Текущее количество строк: 16713
Удалено строк (абсолютное): 243
Удалено строк (относительное): 1.43%


Были удалены как явные, так и не явные дубликат, общее количество - 243.
Потери данных минимальны 1.43%, что незначительно для дальнейшего анализа.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [122]:
import warnings
warnings.filterwarnings('ignore')

In [123]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]

In [124]:
df_actual['user_score_numeric'] = pd.to_numeric(df_actual['user_score'], errors='coerce')

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [125]:
def categorize_user(score):
    if pd.isna(score):
        return None
    elif score >= 8 and score <= 10:
        return 'высокая'
    elif score >= 3 and score < 8:
        return 'средняя'
    elif score >= 0 and score < 3:
        return 'низкая'

df_actual['user_category'] = df_actual['user_score_numeric'].apply(categorize_user)
display(df_actual['user_category'].value_counts(dropna=False))

низкая     6414
средняя    4081
высокая    2286
Name: user_category, dtype: int64

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [126]:
def categorize_critic(score):
    if 80 <= score <= 100:
        return 'высокая'
    elif 30 <= score < 80:
        return 'средняя'
    elif 0 <= score < 30:
        return 'низкая'
    else:
        return 'без оценки'


df_actual['critic_category'] = df_actual['critic_score'].apply(categorize_critic)
display(df_actual['critic_category'].value_counts(dropna=False))


низкая     5667
средняя    5422
высокая    1692
Name: critic_category, dtype: int64

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [127]:
display(df_actual.groupby('user_category')['user_score'].count())

user_category
высокая    2286
низкая     6414
средняя    4081
Name: user_score, dtype: int64

In [128]:
display(df_actual.groupby('critic_category')['user_score'].count())

critic_category
высокая    1692
низкая     5667
средняя    5422
Name: user_score, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [129]:
platform_counts = df_actual['platform'].value_counts()
display(platform_counts)

ps2     2127
ds      2120
wii     1275
psp     1180
x360    1121
ps3     1087
gba      811
xb       803
pc       766
gc       542
3ds      300
ps       274
psv      134
wiiu      74
n64       70
dc        31
gb        27
xone      19
ps4       16
ws         4
Name: platform, dtype: int64

In [130]:
top_7_platforms = platform_counts.head(7)

In [131]:
display(top_7_platforms)

ps2     2127
ds      2120
wii     1275
psp     1180
x360    1121
ps3     1087
gba      811
Name: platform, dtype: int64

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

В ходе выполнения проекта было проделано: подготовка и предобработка данных для дальнейшей корректной работы с ними. Изучив данные, я привел названия к snake_case, преобразовал столбцы eu_sales, jp_sales, user_score в числовой тип, удалил пропуски в name и genre, заполнил пропуски в остальных стобцах, где это требовалось. По условию отфильтровал данные по годам (создан датафрейм df_actual). Были добавлены новые столбцы:'user_category' и 'critic_category', где указаны категории оценок. Далее выделил топ 7 платформ по количеству выпущенных игр.